src/symbols.cpp

```diff
@@ -76,89 +76,193 @@ class MemoryMap {
-            const char* symtab_end = symtab + section->sh_size;
-            for (; symtab < symtab_end; symtab += section->sh_entsize) {
-                Elf64_Sym* sym = (Elf64_Sym*)symtab;
-                if (sym->st_name != 0 && sym->st_value != 0) {
-                    cc->add(base + sym->st_value, (int)sym->st_size, strings + sym->st_name);
-                }
-            }
+void ElfParser::parseMem(CodeCache* cc, const char* base, const void* addr) {
+    ElfParser elf(cc, base, addr);
+    elf.loadSymbols(false);
+}
+
+void ElfParser::loadSymbols(bool use_debug) {
+    if (!valid_header()) {
+        return;
+    }
+
+    // Look for debug symbols in the original .so
+    ElfSection* section = findSection(SHT_SYMTAB, ".symtab");
+    if (section != NULL) {
+        loadSymbolTable(section);
+        return;
+    }
+
+    // Try to load symbols from an external debuginfo library
+    if (use_debug) {
+        if (loadSymbolsUsingBuildId() || loadSymbolsUsingDebugLink()) {
+            return;
         }
     }
 
-    cc->sort();
+    // If everything else fails, load only exported symbols
+    section = findSection(SHT_DYNSYM, ".dynsym");
+    if (section != NULL) {
+        loadSymbolTable(section);
+    }
 }
 
-void Symbols::parseFile(CodeCache* cc, const char* fileName, const char* base) {
-    int fd = open(fileName, O_RDONLY);
-    if (fd != -1) {
-        size_t length = (size_t)lseek64(fd, 0, SEEK_END);
-        char* addr = (char*)mmap(NULL, length, PROT_READ, MAP_PRIVATE, fd, 0);
-        if (addr != NULL) {
-            parseElf(cc, addr, base);
-            munmap(addr, length);
+// Load symbols from /usr/lib/debug/.build-id/ab/cdef1234.debug, where abcdef1234 is Build ID
+bool ElfParser::loadSymbolsUsingBuildId() {
+    ElfSection* section = findSection(SHT_NOTE, ".note.gnu.build-id");
+    if (section == NULL || section->sh_size <= 16) {
+        return false;
+    }
+
+    ElfNote* note = (ElfNote*)at(section);
+    if (note->n_namesz != 4 || note->n_descsz < 2 || note->n_descsz > 64 || note->n_type != NT_GNU_BUILD_ID) {
+        return false;
+    }
+
+    const char* build_id = (const char*)note + sizeof(*note) + 4;
+    int build_id_len = note->n_descsz;
+
+    char path[PATH_MAX];
+    char* p = path + sprintf(path, "/usr/lib/debug/.build-id/%02hhx/", build_id[0]);
+    for (int i = 1; i < build_id_len; i++) {
+        p += sprintf(p, "%02hhx", build_id[i]);
+    }
+    strcpy(p, ".debug");
+
+    return parseFile(_cc, _base, path, false);
+}
+
+// Look for debuginfo file specified in .gnu_debuglink section
+bool ElfParser::loadSymbolsUsingDebugLink() {
+    ElfSection* section = findSection(SHT_PROGBITS, ".gnu_debuglink");
+    if (section == NULL || section->sh_size <= 4) {
+        return false;
+    }
+
+    const char* basename = strrchr(_file_name, '/');
+    if (basename == NULL) {
+        return false;
+    }
+
+    char* dirname = strndup(_file_name, basename - _file_name);
```

Comment:
Yes, I know about these library functions, but I avoided them intentionally.
dirname(3) yields extra '/' for the root directory. Also man page is not quite clear whether it is safe to call dirname+basename on the same string. Since these library calls do not make the code substantially shorter or clearer, I would stay with the given implementation.